### ISSUE TITLE
Allow passing specific depth to `Document::getFeatures`

### ISSUE THREAD

**Issue by MaKleSoft** [Polymer/polymer-analyzer#352]
Offers more control and should be fairly simple to implement. Off the top of my head, something like this should work:

```js
  getFeatures(deep?: bool | number): Set<Feature> {
    if (deep == null) {
      deep = true;
    }
    const result = new Set<Feature>();
    this._getFeatures(result, new Set<Document>(), deep);
    return result;
  }

  private _getFeatures(
      result: Set<Feature>, visited: Set<Document>, deep: boolean | number) {
    if (visited.has(this)) {
      return;
    }
    visited.add(this);
    for (const feature of this._localFeatures) {
      result.add(feature);
      if (deep) {
        const depth = typeof deep === "number" ? deep - 1 : deep;
        if (feature.kinds.has('document')) {
          (feature as Document)._getFeatures(result, visited, depth);
        }
        if (feature.kinds.has('import')) {
          (feature as Import).document._getFeatures(result, visited, depth);
        }
      }
    }
  }
```
If this is something you want to support I'll be happy to create a pull request.

**Comment by rictic** [comment]
/cc @justinfagnani who's been thinking about revamping these methods.

It might be overly clever, but if we support passing `depth` as a number, we could represent the current `deep == true` behavior by setting `depth` to `Infinity` :)

**Comment by rictic** [comment]
Do you have a use case for this feature request? For a lot of cases I'm thinking about you really only care about a depth of 1 and a depth of Infinity. e.g. when do you care about whether an element was defined in an import vs an import of an import?

I suppose it would be cool to be able to warn on using elements that weren't defined in a document or directly in one of its imports. But this is slightly tricky because `<script src>` is also treated as an import, so I don't think `depth` would be especially helpful here. Though I suppose just setting `depth` to 3 would more or less do what you want. Hm.

**Comment by MaKleSoft** [comment]
> It might be overly clever, but if we support passing depth as a number, we could represent the current deep == true behavior by setting depth to Infinity :)

Haha, I was thinking the same thing. The nice thing is that `Infinity - 1 == Infinity`. I agree that it might be a little too clever though :)

> Do you have a use case for this feature request?

Exactly this:
> I suppose it would be cool to be able to warn on using elements that weren't defined in a document or directly in one of its imports.

I think this wouldn't be just cool, but incredibly valuable. Polylint currently only checks if all used elements are defined __somewhere__, which is is not optimal since indirectly imported elements can easily lead to fragile import chains. Lets say element __a__ uses element __b__ and element __c__ but only imports __b__ directly. In practice this is not a problem since __b__ happens to be importing __c__. What happens if during a refactor, __b__ is no longer used by __a__, but __c__ still is? The app is breaks, because no definition for __c__ is being loaded anymore. I think this is a very common pitfall, so it would be nice to be able to identify this kind of issue easily.

**Comment by rictic** [comment]
I strongly agree that that lint rule would be supremely useful.

If that's the only use case though, it's probably better to just do the `depth = 3` logic inside of that one lint rule rather than making the query APIs more complex. It's often the case that lint rules need to reimplement stuff or do somewhat weird things that you really don't need to do anywhere else. For example we're implementing the search for polymer expression delimiters ( `{{...}}` and `[[...]]`) in two places, once in the analyzer, and then again in a linter, where it's looking for incorrect delimiters (`[...]]`, `{{...]]`, etc).

**Comment by MaKleSoft** [comment]
Would this make the query api that much more complex though? The way I see it we gain a lot of flexibility with only minimal added complexity.

**Comment by rictic** [comment]
Because a bunch of different tools are using the analyzer we're striving really hard to make it performant. The `getBy*` methods should be as fast as possible, and with just `deep = false` and `deep = true` it's reasonably easy to make them constant-time lookups in maps. If we support arbitrary depths then we've either got a much more complex indexing story, or we've got a hidden performance footgun (granted, one that very few callers are likely to need or use).

It's also just a bit more confusing. If I want only features in the document I'm holding, do I say `depth = 1` or `depth = 0` or `depth = false`?

All of this is manageable complexity, and I think it's worth it if we run into a couple more use cases. But I want to be deliberate about adding anything to document, as it's the primary API surface of the analyzer, and any complexity that we add there could be hard to walk back without breaking users.

**Comment by MaKleSoft** [comment]
Yeah, I wasn't considering performance. To be honest I can't think of any other use cases right now, either, but I have a gut feeling that they are out there...

**Comment by rictic** [comment]
I hear you. Let's leave this issue open for a while and see if more use cases come out of the woodwork. As your initial proposal shows, it would be pretty easy to go from `boolean` to `boolean | number`.

**Comment by justinfagnani** [comment]
I'm not sure how one would be expected to know what depth to pass. What would non-trivial values even mean in a complex graph?

The cases we care about are: 1) a document defining something locally 2) a document importing something 3) a document exporting something.

In a scoped world, deep wouldn't even matter, because you don't see what you don't directly import. The only reason why we even think of deep is because so many things are global, and you can see things imported several steps away.

I'd rather move towards thinking in scopes first, and thinking of the global scopes (JS and custom elements are separate global scopes) as ones that each document has and that automatically re-exports everything they import. Then we don't think of deep, but local, imported and exported.

It definitely would be nice to warn on not directly importing your dependencies, but it's trickier than just limiting depth to 1. HTML Imports don't have a re-export mechanism, and some packages are purposefully organized to have a document that acts as a re-export. How could we tell the difference between this and an accidental use of a transitive dependency? I think we can do it with some convention - maybe we put a comment on an import that indicates to tools that it's a "re-export". THen we don't need the depth parameter and can differentiate between the two cases.

**Comment by MaKleSoft** [comment]
@justinfagnani You're making some very good points. Seems like I was underestimating the complexity of the issue...
